src/client/datagram.cc

```diff
@@ -52,6 +52,41 @@ DatagramDestination::DatagramDestination(
     : m_Owner(owner),
       m_Receiver(nullptr) {}
 
+namespace {  // Helper facilities to transfer messages.
+
+/// Deleter for a shared pointer
+/// that acts like double handle.
+template <typename T>
+struct DeepDeleter {
+  /// Deletes the payload and handle.
+  void operator()(T** ptr) {
+    delete *ptr;
+    delete ptr;
+  }
+};
+
+/// Shared pointer with ``release`` semantics as unique_ptr.
+template <typename T>
+using ReleasableSharedPtr = std::shared_ptr<T*>;
+
+/// Helper function to make releasable shared_ptr with appropriate deleter.
+//
+/// @param ptr  The payload for the handle.
+template <typename T>
+auto make_releasable_shared_ptr(T* payload) {
+  return ReleasableSharedPtr<T>(new T*(payload), DeepDeleter<T>());
+}
+
+/// @returns The released payload from the shared ptr.
+template <typename T>
+T* release(ReleasableSharedPtr<T>& smart_ptr) {
+  T* paylaod = *smart_ptr;
+  *smart_ptr = nullptr;
+  return paylaod;
+}
+
+}  // namespace
+
```

Comment:
Looking at ```datagram``` code, I can tell that I've spent almost no time in this area (its hideous), so I can't give any solid opinions at the moment other than:

1. If you're willing to dive deeper, refactor bigger design to accommodate true ```unique_ptr```
2. If diving deeper, you see that ```shared_ptr``` is the only option (I don't see why though (who else *really* needs ownership of the message?)), then feel free to refactor
3. Or, add a huge ```TODO``` to resolve this issue and we will merge this and get back to it at another time

Or maybe some other ideas?